flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/Scheduler.java

```diff
@@ -0,0 +1,516 @@
+
+		SlotRequestId slotRequestId = logicalSlot.getSlotRequestId();
+		SlotSharingGroupId slotSharingGroupId = logicalSlot.getSlotSharingGroupId();
+		FlinkException cause = new FlinkException("Slot is being returned to the SlotPool.");
+
+		cancelSlotRequest(slotRequestId, slotSharingGroupId, cause);
+		return true;
+	}
+
+	//---------------------------
+
+	private CompletableFuture<LogicalSlot> allocateSingleSlot(
+		SlotRequestId slotRequestId,
+		SlotProfile slotProfile,
+		boolean allowQueuedScheduling,
+		Time allocationTimeout) {
+
+		SlotAndLocality slotAndLocality =
+			tryAllocateFromAvailable(slotRequestId, slotProfile);
+
+		if (slotAndLocality != null) {
+			// already successful from available
+			return CompletableFuture.completedFuture(
+				completeAllocationByAssigningPayload(slotRequestId, slotAndLocality));
+		} else if (allowQueuedScheduling) {
+			// we allocate by requesting a new slot
+			return slotPoolGateway
+				.requestNewAllocatedSlot(slotRequestId, slotProfile.getResourceProfile(), allocationTimeout)
+				.thenApply((AllocatedSlot allocatedSlot) ->
+					completeAllocationByAssigningPayload(slotRequestId, new SlotAndLocality(allocatedSlot, Locality.UNKNOWN)));
+		} else {
+			// failed to allocate
+			return FutureUtils.completedExceptionally(
+				new NoResourceAvailableException("Could not allocate a simple slot for " + slotRequestId + '.'));
+		}
+	}
+
+	@Nonnull
+	private LogicalSlot completeAllocationByAssigningPayload(
+		@Nonnull SlotRequestId slotRequestId,
+		@Nonnull SlotAndLocality slotAndLocality) {
+
+		final AllocatedSlot allocatedSlot = slotAndLocality.getSlot();
+
+		final SingleLogicalSlot singleTaskSlot = new SingleLogicalSlot(
+			slotRequestId,
+			allocatedSlot,
+			null,
+			slotAndLocality.getLocality(),
+			this);
+
+		if (allocatedSlot.tryAssignPayload(singleTaskSlot)) {
+			return singleTaskSlot;
+		} else {
+			final FlinkException flinkException =
+				new FlinkException("Could not assign payload to allocated slot " + allocatedSlot.getAllocationId() + '.');
+			slotPoolGateway.releaseSlot(slotRequestId, flinkException);
+			throw new CompletionException(flinkException);
+		}
+	}
+
+	private SlotAndLocality tryAllocateFromAvailable(
+		@Nonnull SlotRequestId slotRequestId,
+		@Nonnull SlotProfile slotProfile) {
+
+		List<SlotInfo> slotInfoList = slotPoolGateway.getAvailableSlotsInformation();
+
+		SlotSelectionStrategy.SlotInfoAndLocality selectedAvailableSlot =
+			slotSelectionStrategy.selectBestSlotForProfile(slotInfoList, slotProfile);
+
+		SlotInfo selectedSlotInfo = selectedAvailableSlot.getSlotInfo();
+
+		if (selectedSlotInfo != null) {
+			AllocatedSlot allocatedSlot =
+				slotPoolGateway.allocateAvailableSlot(slotRequestId, selectedSlotInfo.getAllocationId());
+			if (allocatedSlot != null) {
+				return new SlotAndLocality(allocatedSlot, selectedAvailableSlot.getLocality());
+			}
+		}
+
+		// no available slot fits
+		return null;
```

Comment:
Either add `@Nullable` to the return type or slightly better use `Optional` to indicate that this method might fail.